### PR TITLE
Archive viewer

### PR DESCRIPTION
**Pull request by hius07**
Sometimes I am curious about the archive content.
Zip (epub, cbz) only so far.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10584)

### CODE REVIEW

File: plugins/arcviewer.koplugin/main.lua [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            end
+        else
+            if self.list_table[path][name] == nil then
+                self.list_table[path][name] = filesize
+                parse_path(path:sub(1,-2), false) -- go up to include subfolders of the branch into the table
+            end
+        end
+    end
+
+    local std_out = io.popen("unzip ".."-qql \""..file.."\"")
```
Comment (NiLuJe) [inline plugins/arcviewer.koplugin/main.lua:93]: Can we do that on Android?

File: plugins/arcviewer.koplugin/_meta.lua [hunk 2]
```diff
@@ -0,0 +1,6 @@
+local _ = require("gettext")
+return {
+    name = "arcviewer",
+    fullname = _("Archive viewer"),
+    description = _([[Archives viewer.]]),
```
Comment (Frenzie) [inline plugins/arcviewer.koplugin/_meta.lua:5]:
```suggestion
    description = _([[Enables exploring the content of zip archives.]]),
```

File: plugins/arcviewer.koplugin/_meta.lua [hunk 3]
```diff
@@ -0,0 +1,6 @@
+local _ = require("gettext")
+return {
+    name = "arcviewer",
```
Comment (poire-z) [inline plugins/arcviewer.koplugin/_meta.lua:3]: "archiveviewer" is not too long, no need to make it that concise - and "arcviewer" to me sounds like "arc format viewer".

File: plugins/arcviewer.koplugin/main.lua [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+local Screen = require("device").screen
+local T = ffiUtil.template
+
+local ArcViewer = WidgetContainer:extend{
+    -- list_table is a flat table containing archive files and folders
+    -- key - a full path of the folder ("/" for root), for all folders and subfolders of any level
+    -- value - a subtable of subfolders and files in the folder
+    -- subtable key - a name of a subfolder ending with /, or a name of a file (without path)
+    -- subtable value - false for subfolders, or file size (string)
+    list_tabe = nil,
```
Comment (poire-z) [inline plugins/arcviewer.koplugin/main.lua:19]: typo (`list_tabLe`)

File: plugins/arcviewer.koplugin/_meta.lua (hunk 3 above)
Comment (poire-z) [reply plugins/arcviewer.koplugin/_meta.lua]: The plugin subdirectory should also be renamed from arcviewer.koplugin to archiveviewer.koplugin.
(Dunno if you can do that from the web interface?)

File: plugins/arcviewer.koplugin/main.lua [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        epub = true,
+        zip  = true,
+    },
+}
+
+function ArcViewer:isSupported(file)
+    return self.arc_ext[util.getFileNameSuffix(file):lower()] and true or false
+end
+
+function ArcViewer:openArcViewer(file)
```
Comment (poire-z) [inline plugins/arcviewer.koplugin/main.lua:32]: :openArchiveViewer (or :viewArchive)

File: plugins/arcviewer.koplugin/main.lua [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+local Menu = require("ui/widget/menu")
+local UIManager = require("ui/uimanager")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local ffiUtil = require("ffi/util")
+local util = require("util")
+local _ = require("gettext")
+local Screen = require("device").screen
+local T = ffiUtil.template
+
+local ArcViewer = WidgetContainer:extend{
```
Comment (poire-z) [inline plugins/arcviewer.koplugin/main.lua:13]: local ArchiveViewer

File: plugins/arcviewer.koplugin/main.lua [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        text = T("%1 \u{F114} ", sub_dirs) .. text
+    end
+    return text
+end
+
+function ArcViewer:extractFile(arcfile, filepath)
+    UIManager:show(ConfirmBox:new{
+        text = _("Extract this file?") .. "\n\n" .. filepath .. "\n\n" ..
+            _("If the file already exists, it will be overwritten."),
+        ok_text = _("Extract"),
```
Comment (poire-z) [inline plugins/arcviewer.koplugin/main.lua:175]: This plugins only extract files ? and we then have to exit it and go browse to the extraction location to then view the file ?
It is a bit lying to its name :) I would be expecting to view an image in ImageViewer, and any other stuff in TextViewer.

**Review by poire-z — APPROVED** [review]

**Comment by poire-z** [comment]
Upgraded KOReader today - tried this plugin, but it always shows me "no items" (even on really simple epub filenames without any space or problematic chars).
On my Kobo and on the emulator.

**Comment by hius07** [comment]
Does `unzip` command in Terminal emulator work?

**Comment by poire-z** [comment]
unzip is found, and replies. But may be not in the format you expect.

The output of my debian unzip (UnZip 6.00 of 20 April 2009, by Debian) on the emulator:
(added some print(line), print(fsize), print(fname)
```
    11273  2015-10-07 23:20   text/part0069_split_015.html
nil
nil
      797  2015-10-07 23:20   text/part0070.html
nil
nil
```
And on my Kobo:
```
     2101  05-23-17 06:06   OEBPS/titl001.xhtml
nil
nil
      928  05-23-17 06:06   OEBPS/title.css
nil
nil
    59818  05-23-17 06:06   OEBPS/toc.ncx
nil
nil
```

Your `local fsize, fname = string.match(line, "%s+(%d+)%s+%d%d%-%d%d%-%d%d%d%d%s+%d%d:%d%d%s+(.+)")` should probably be changed to match these 3 kinds of date time (hoping than on the other devices, only it can change too.)
This seems to work on both the emulator and my (old firmware) Kobo:
`local fsize, fname = string.match(line, "%s+(%d+)%s+[-0-9]+%s+[0-9:]+%s+(.+)")`
Works on your Kindle ?

**Pull request merged**